apps/prairielearn/src/pages/instructorAssessmentDownloads/instructorAssessmentDownloads.ts

```diff
@@ -346,7 +345,7 @@ router.get(
 
       for await (const rows of cursor.iterate(100)) {
         for (const row of rows) {
-          const contents = row.contents != null ? row.contents : '';
+          const contents = Buffer.from(row.contents ?? '', 'base64');
```

Comment:
AFAICT there's nothing in our schema that prevents `contents` from being an arbitrary JSON value. `Buffer.from` will throw if the first argument is anything other than a string.

`base64_safe_decode` would return `NULL` if decoding failed. I think we should do the same thing here: in the end, if a file's `contents` is not a string, add it as an empty file.

We could consider adding some robustness around `row.filename` as well:

- Ignore the file if `filename` isn't a string
- Ignore the file if `filename` tries to do directory traversal, maybe?

Reply:
> AFAICT there's nothing in our schema that prevents `contents` from being an arbitrary JSON value. `Buffer.from` will throw if the first argument is anything other than a string.

Good point. I was considering a null/missing `contents` or invalid encoding, I didn't consider a different type.

> `base64_safe_decode` would return `NULL` if decoding failed. I think we should do the same thing here: in the end, if a file's `contents` is not a string, add it as an empty file.

I'll create a wrapper.

> We could consider adding some robustness around `row.filename` as well:
> 
> * Ignore the file if `filename` isn't a string

Will do.

> * Ignore the file if `filename` tries to do directory traversal, maybe?

This can be a valid case for questions with workspaces, where the graded files may be inside directories.

Reply:
By "directory traversal" I mean https://owasp.org/www-community/attacks/Path_Traversal. Essentially, a filename could contain `../` and (over)write unrelated files in the archive. Granted, the filenames should all be instructor-controlled, so this would really only be a problem if instructors do something really stupid... so maybe it's not worth worrying about.

Reply:
I'm not concerned about errors per se; I just think we shouldn't blindly assume that arrays/objects/etc. can and should be used to construct a Buffer. The PL convention is that files are base64-encoded strings and IMO we should enforce that here.

Reply:
Thanks for pointing out that the query will already produce a string filename; we can leave the JS handling the same given that.

Reply:
> I'm not concerned about errors per se; I just think we shouldn't blindly assume that arrays/objects/etc. can and should be used to construct a Buffer. The PL convention is that files are base64-encoded strings and IMO we should enforce that here.

Done in e4f8dc949fefe7e283ad950a8a194936b722031a.